torch/csrc/jit/interpreter.cpp

```diff
@@ -384,35 +390,89 @@ Operation createPythonOperation(PythonOp* op) {
     size_t next_scalar = 0;
     size_t next_tensor = 0;
     HandleBuilder builder(has_handle);
-    for(auto arg_type : op->cconv) {
-      if(arg_type == 's') {
-        py_inputs[i] = py::reinterpret_borrow<py::object>(op->scalar_args[next_scalar++].get());
-      } else if(arg_type == 't') {
-        py_inputs[i] = py::reinterpret_steal<py::object>(THPVariable_Wrap(
-          builder.addInput(std::move(fromLast(stack, num_inputs - next_tensor)), op->var_flags.at(next_tensor))));
-        next_tensor++;
+    if (unpack_variables) {
+      for (auto arg_type : op->cconv) {
+        if (arg_type == 's') {
+          py_inputs[i] = py::reinterpret_borrow<py::object>(
+              op->scalar_args[next_scalar++].get());
+        } else if (arg_type == 't') {
+          py_inputs[i] = py::reinterpret_steal<py::object>(
+              THPVariable_Wrap(builder.addInput(
+                  std::move(fromLast(stack, num_inputs - next_tensor)),
+                  op->var_flags.at(next_tensor))));
```

Comment:
Wait, does `unpack_variables` mean that you want to wrap into variables? I'm confused how this interacts with `values_are_variables`

Reply:
The names for the flags are weird right now `unpack_variables` should probably be called `tracing_autograd_python_function`. Then `values_are_variables` determines if the inputs to the python function are already variables or if they are tensors. Because of optimization passes, it is possible that the graph executor has decided to unwrap variables to do more aggressive optimizations. In that case, there may be a call to a python op inside of one of these blocks, and we can see `values_are_variables==false`. This requires wrapping/unwrapping around the call.

Reply:
How is this possible? PythonOps are not differentiable/have no type info and therefore will never be part of white box graphs, and those are the only places  where we unwrap the tensors. I'd rather make the code simpler if we can, because right now it's quite bad.

Reply:
Right, good point. I still think it would be beneficial to unify the arg packing/unpacking logic, because right now it's a bit hard to find the differences.

Reply:
Or, if there's no easy way of doing this, we should list the differences in a comment before the two branches

Reply:
I am not sure the logic is actually the same across the two cases (autograd tracing vs forward-only tracing). In the first case variables get hooked up to handles.